Source/Core/DolphinQt2/Debugger/MemoryWidget.cpp

```diff
@@ -0,0 +1,576 @@
+  SaveSettings();
+}
+
+void MemoryWidget::OnBPTypeChanged()
+{
+  bool read_write = m_bp_read_write->isChecked();
+  bool read_only = m_bp_read_only->isChecked();
+
+  MemoryViewWidget::BPType type;
+
+  if (read_write)
+    type = MemoryViewWidget::BPType::ReadWrite;
+  else if (read_only)
+    type = MemoryViewWidget::BPType::ReadOnly;
+  else
+    type = MemoryViewWidget::BPType::WriteOnly;
+
+  m_memory_view->SetBPType(type);
+
+  SaveSettings();
+}
+
+void MemoryWidget::OnSearchAddress()
+{
+  bool good;
+  u32 addr = m_search_address->text().toUInt(&good, 16);
+
+  QFont font;
+  QPalette palette;
+
+  if (good || m_search_address->text().isEmpty())
+  {
+    if (good)
+      m_memory_view->SetAddress(addr);
+  }
+  else
+  {
+    font.setBold(true);
+    palette.setColor(QPalette::Text, Qt::red);
+  }
+
+  m_search_address->setFont(font);
+  m_search_address->setPalette(palette);
+}
+
+void MemoryWidget::ValidateSearchValue()
+{
+  QFont font;
+  QPalette palette;
+
+  if (m_find_hex->isChecked() && !m_data_edit->text().isEmpty())
+  {
+    bool good;
+    m_data_edit->text().toULongLong(&good, 16);
+
+    if (!good)
+    {
+      font.setBold(true);
+      palette.setColor(QPalette::Text, Qt::red);
+    }
+  }
+
+  m_data_edit->setFont(font);
+  m_data_edit->setPalette(palette);
+}
+
+void MemoryWidget::OnSetValue()
+{
+  bool good_address;
+  u32 addr = m_search_address->text().toUInt(&good_address, 16);
+
+  if (!good_address)
+  {
+    QMessageBox::critical(this, tr("Error"), tr("Bad address provided."));
+    return;
+  }
+
+  if (m_data_edit->text().isEmpty())
+  {
+    QMessageBox::critical(this, tr("Error"), tr("No value provided."));
+    return;
+  }
+
+  if (m_find_ascii->isChecked())
+  {
+    std::string ascii = m_data_edit->text().toStdString();
+
+    for (char c : ascii)
+      PowerPC::HostWrite_U8(static_cast<u8>(c), addr++);
+  }
+  else
+  {
+    bool good_value;
+    u64 value = m_data_edit->text().toULongLong(&good_value, 16);
+
+    if (!good_value)
+    {
+      QMessageBox::critical(this, tr("Error"), tr("Bad value provided."));
+      return;
+    }
+
+    if (value == static_cast<u8>(value))
+    {
+      PowerPC::HostWrite_U8(static_cast<u8>(value), addr);
+    }
+    else if (value == static_cast<u16>(value))
+    {
+      PowerPC::HostWrite_U16(static_cast<u16>(value), addr);
+    }
+    else if (value == static_cast<u32>(value))
+    {
+      PowerPC::HostWrite_U32(static_cast<u32>(value), addr);
+    }
+    else
+      PowerPC::HostWrite_U64(value, addr);
+  }
+
+  Update();
+}
+
+static void DumpArray(const std::string& filename, const u8* data, size_t length)
+{
+  if (!data)
+    return;
+
+  File::IOFile f(filename, "wb");
+
+  if (!f)
+  {
+    QMessageBox::critical(
+        nullptr, QObject::tr("Error"),
+        QObject::tr("Failed to dump %1: Can't open file").arg(QString::fromStdString(filename)));
+    return;
+  }
+
+  f.WriteBytes(data, length);
```

Comment:
It looks like you forgot to handle write failures too.

Reply:
Done.